server/channels/jobs/batch_report_worker.go

```diff
@@ -0,0 +1,149 @@
+		stopped:            make(chan bool, 1),
+		jobs:               make(chan model.Job),
+		timeBetweenBatches: timeBetweenBatches,
+		doBatch:            worker.doBatch,
+	}
+	return worker
+}
+
+func (worker *BatchReportWorker[T]) doBatch(rctx *request.Context, job *model.Job) bool {
+	reportData, nextData, done, err := worker.getData(job.Data, worker.app)
+	if err != nil {
+		worker.logger.Error("Worker: Failed to get data for report batch. Exiting", mlog.Err(err))
+		worker.setJobError(worker.logger, job, model.NewAppError("doBatch", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+		return true
+	} else if done {
+		if err = worker.complete(rctx, job); err != nil {
+			worker.logger.Error("Worker: Failed to finish the batch report. Exiting", mlog.Err(err))
+			worker.setJobError(worker.logger, job, model.NewAppError("doBatch", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+		} else {
+			worker.logger.Info("Worker: Report job complete")
+			worker.setJobSuccess(worker.logger, job)
+		}
+
+		return true
+	}
+
+	err = worker.saveData(job, reportData)
+	if err != nil {
+		worker.logger.Error("Worker: Failed to save report batch. Exiting", mlog.Err(err))
+		worker.setJobError(worker.logger, job, model.NewAppError("doBatch", model.NoTranslation, nil, "", http.StatusInternalServerError).Wrap(err))
+		return true
+	}
+
+	for k, v := range nextData {
+		job.Data[k] = v
+	}
+
+	// TODO add progress?
+	worker.jobServer.SetJobProgress(job, 0)
+	return false
+}
+
+func getFileCount(jobData model.StringMap) (int, error) {
+	if jobData["file_count"] != "" {
+		parsedFileCount, parseErr := strconv.Atoi(jobData["file_count"])
+		if parseErr != nil {
+			return 0, errors.Wrap(parseErr, "failed to parse file_count")
+		}
+		return parsedFileCount, nil
+	}
+
+	// Assume it hasn't been set
+	return 0, nil
+}
+
+func (worker *BatchReportWorker[T]) saveData(job *model.Job, reportData []model.ReportableObject) error {
+	fileCount, err := getFileCount(job.Data)
+	if err != nil {
+		return err
+	}
+
+	appErr := worker.app.SaveReportChunk(worker.reportFormat, job.Id, fileCount, reportData)
+	if appErr != nil {
+		return err
+	}
+
+	fileCount++
+	job.Data["file_count"] = strconv.Itoa(fileCount)
+
+	return nil
+}
+
+func (worker *BatchReportWorker[T]) complete(rctx request.CTX, job *model.Job) error {
+	requestingUserId := job.Data["requesting_user_id"]
+	if requestingUserId == "" {
+		return errors.New("No user to send the report to")
+	}
+	fileCount, err := getFileCount(job.Data)
+	if err != nil {
+		return err
+	}
+
+	appErr := worker.app.CompileReportChunks(worker.reportFormat, job.Id, fileCount, worker.headers)
+	if appErr != nil {
+		return appErr
+	}
+
+	if appErr = worker.app.SendReportToUser(rctx, requestingUserId, job.Id, worker.reportFormat); appErr != nil {
+		return appErr
+	}
+
+	go func() {
+		worker.app.CleanupReportChunks(worker.reportFormat, job.Id, fileCount)
+	}()
```

Comment:
If we fail to send above, will we ever clean up these files?

Reply:
Good point. This should probably be a deferred function that we add after the compilation.
One thing to maybe consider going forward is a method of cleaning out the file store if there are other failed reports.